text/0019-custom-shaders.md

```diff
@@ -0,0 +1,299 @@
+``` 
+
+We define a struct called `StarMoon` that implements the `Widget` trait, and contains the following fields:
+
+```rs
+    pub struct StarMoon {
+        size: f32,
+        mouse_click: Vector,
+        duration_since_start: Duration,
+    }
+```
+
+
+In similar fashion to custom quads, the `draw(..)` method for the `Widget` implementation of `StarMoon` is written as follows:
+```rs
+        fn draw(
+            &self,
+            _state: &widget::Tree,
+            renderer: &mut Renderer,
+            _theme: &Renderer::Theme,
+            _style: &renderer::Style,
+            layout: Layout<'_>,
+            cursor_position: Point,
+            _viewport: &Rectangle,
+        ) {
+            renderer.make_custom_shader_quad(
+                renderer::CustomShaderQuad {
+                    bounds: layout.bounds(),
+                    
+                    mouse_position: cursor_position,
+                    mouse_click: self.mouse_click,
+                    time: self.duration_since_start.as_secs_f32(),
+                    frame_number: 1,
+
+                    shader_code: SHADER.to_string(),
+                },
+                Color::from_rgb(1.0, 0.0, 0.0),
+            );
+        }
+```
+
+On the shader side, the information is received in the form of attributes:
+
+```rs
+struct VertexInput {
+    @location(0) v_pos: vec2<f32>,
+    @location(1) pos: vec2<f32>,
+    @location(2) size: vec2<f32>,
+    @location(3) bg_color: vec4<f32>,
+
+    @location(4) mouse_position: vec2<f32>,
+    @location(5) mouse_click: vec2<f32>,
+    @location(6) time: f32,
+    @location(7) frame: u32,
+}
+```
+
+The mouse clicks are encoded using a two-dimensional vector: the first component corresponds to a left mouse button press, and the second component to a right mouse button press. In this particular example, their values are zero at rest and one when pressed. In the shader code (see `custom_shader.wgsl` in the example), the color of the shapes react to a right mouse button press:
+    
+```rs
+    // upon left mouse button press, change border color
+    if input.mouse_click.y > 0.5 {
+        shape_color = vec4<f32>(0.84, 0.05, 0.92, 1.); // purple
+    }
+
+```
+
+
+<br/><br/> 
+## Implementation strategy
+
+See the actual implementation in my fork of Iced: https://github.com/eliotbo/iced.
+
+The implementation of custom shader quads should follow that of custom quads as closely as possible. In summary, we modify the following modules: `graphics::renderer`, `wgpu::backend`, `graphics::primitive`, `graphics::layer` and `native::renderer`. And we add two new modules: `graphics::layer::custom_shader_quad` and `wgpu::custom_shader_quad`.
+
+We add a new  `graphics::Primitive`:
+```rs
+    CustomShaderQuad {
+        bounds: Rectangle,
+        background: Background, 
+        mouse_position: Vector<f32>,
+        mouse_click: Vector<f32>,
+        time: f32,
+        frame: u32,
+        shader_code: String,
```

Comment:
Passing the shader as a string in the primitive will not only require an allocation every time Iced generates a new frame with new primitives, but also require recompiling the shader, right? Probably should use something reference counted that either contains the compiled shader or allows it to be cached by the backend, like `image::Handle`.

Reply:
Passing a handle to the shader code is the right way to go, but I'm not sure how to implement it. I'm wondering if simply changing the type from `String` to `&str` could work, although I am sure we would run into countless lifetime issues.

For reference in the the case of `wgpu::quad`, the pipeline inserts the shader like so:

```rs
        let shader =
            device.create_shader_module(wgpu::ShaderModuleDescriptor {
                label: Some("iced_wgpu::quad::shader"),
                source: wgpu::ShaderSource::Wgsl(std::borrow::Cow::Borrowed(
                    include_str!("shader/quad.wgsl"),
                )),
            });
```

while the pipeline for `custom_shader_quad` does this:
```rs
        let shader =
            device.create_shader_module(wgpu::ShaderModuleDescriptor {
                label: Some("iced_wgpu::custom shader quad::shader"),
                source: wgpu::ShaderSource::Wgsl(std::borrow::Cow::Borrowed(
                    shader,
                )),
            });
```
where the `shader` argument is a `&str`. Are the shaders not compiled at every frame in both cases?

Reply:
That `create_shader_module()` call in `quad.rs` is in `Pipeline::new`, not `Pipeline::draw`. So it's not called every draw and only when the backend is created.

Reply:
I've had a look at image::Handle and I think I've come up with a solution to both the memory allocation issue and the shader compilation problem. Here it goes.

There is a new `shader.rs` module in `iced_native` with a `Handle` struct:
```rs
pub struct Handle {
    /// A unique identifier for the shader.
    pub id: u64,
    /// The path to the shader code.
    pub path: PathBuf,
}
```
Instead of passing the shader as a string in the primitive, we pass the absolute path to the shader file. Internally, a `iced_native::shader::Handle` is created with a hashed `id` based on said absolute path. 

In the `wgpu::custom_shader_quad` `Pipeline`, there are two new fields:
```rs
pub struct Pipeline {
    ...
    pipeline: wgpu::RenderPipeline,
    shader_modules_cache: HashMap<u64, wgpu::ShaderModule>,
}
```
where `shader_modules_cache` is used to cache the compiled `ShaderModule`s. For each `ShaderModule`, the key to the `HashMap` is the corresponding `shader::Handle` id. 

In the `draw(..)` call, the `shader_modules_cache` and the `Pipeline` are updated only if the received `Handle` id is brand new:
```rs
            let shader_handle: Handle = instances[i].shader_handle.clone();

            let has_new_shader_module =
                self.insert_shader_module(device, &shader_handle);

            if has_new_shader_module {
                let shader_module = self
                    .shader_modules_cache
                    .get(&shader_handle.id)
                    .unwrap()
                    .clone();

                let new_pipeline = self.make_pipeline(&device, shader_module);

                self.pipeline = new_pipeline;
```

Reply:
With image handles, you can load an image either from a path or for memory. It may make sense to do the same.

Reply:
I've changed the `native::shader` module:

```rs
pub struct Handle {
    /// A unique identifier for the shader.
    pub id: u64,
    /// Either the path to the shader code or a reference to the shader code in memory.
    pub shader_content: ShaderContent,
}

pub enum ShaderContent {
    /// Shader in a file
    Path(PathBuf),
    /// Shader in memory
    Memory(&'static str),
}
```